### PR TITLE
for #271: allow visitors to join the waitlist

### PR DESCRIPTION
**Pull request by groovecoder**
Adds a `WAITLIST_OPEN` env var to allow visitors to "join the waitlist" by creating an inactive `Invitation` object for them.

Later, we can set the Invitations to `active=True` to invite them.

TODO:

* Needs better UI in the waitlist template. Something like we had on the Monitor page would probably work.

### CODE REVIEW

File: privaterelay/signals.py [hunk 1]
```diff
@@ -22,6 +22,12 @@ def invitations_only(sender, **kwargs):
         if domain_part == allowed_domain:
             return True
 
+    # TODO: put sociallogin in session to validate?
+    kwargs['request'].session['waitlist_email'] = email
+    kwargs['request'].session['waitlist_avatar'] = (
+        sociallogin.account.extra_data['avatar']
+    )
```
Comment (skim1102) [inline privaterelay/signals.py:29]:
```suggestion
    if settings.WAITLIST_OPEN:
        kwargs['request'].session['waitlist_open'] = True
        kwargs['request'].session['waitlist_email'] = email
        kwargs['request'].session['waitlist_avatar'] = (
            sociallogin.account.extra_data['avatar']
        )
    else:
        kwargs['request'].session['waitlist_open'] = False
```

File: privaterelay/templates/socialaccount/authentication_error.html [hunk 2]
```diff
@@ -7,7 +7,14 @@
...
         {% else %}
-            <p class="ff-Met no-active-aliases text-center">Sorry, the Private Relay Beta is currently by invitation only. Please check again later.</p>
+            <p class="ff-Met no-active-aliases text-center">Sorry, the Private Relay Beta is currently by invitation only.</p>
+
+            <form method="post" action="{% url 'waitlist' %}">
+                {% csrf_token %}
+                <input type="hidden" name="email" type="hidden" value="{{ request.session.waitlist_email }}">
+                <img src="{{ request.session.waitlist_avatar }}">
+                <input type="submit" value="Join the Waitlist">
+            </form>
```
Comment (skim1102) [inline privaterelay/templates/socialaccount/authentication_error.html:17]:
```suggestion
        {% if request.session.waitlist_open %}
                <form method="post" action="{% url 'waitlist' %}">
                    {% csrf_token %}
                    <input type="hidden" name="email" type="hidden" value="{{ request.session.waitlist_email }}">
                    <img src="{{ request.session.waitlist_avatar }}">
                    <input type="submit" value="Join the Waitlist">
                </form>
        {% endif %}
```

File: privaterelay/views.py [hunk 3]
```diff
@@ -117,6 +119,19 @@ def invitation(request):
     return redirect('/')
 
 
+@require_http_methods(['POST'])
+def waitlist(request):
+    email = request.POST.get('email')
```
Comment (skim1102) [inline privaterelay/views.py:124]:
```suggestion
    if not settings.WAITLIST_OPEN:
        raise PermissionDenied
    email = request.POST.get('email')
```
Prevent invite creation when waitlist is not open.

File: static/js/app.js [hunk 4]
```diff
@@ -152,6 +152,19 @@ function toggleVisibilityOfElementsIfAddonIsInstalled() {
...
+async function addEmailToWaitlist(e) {
+  e.preventDefault();
+  const waitlistForm = e.target;
+  const formData = {
+    "csrfmiddlewaretoken": waitlistForm[0].value,
+    "email": waitlistForm[1].value
+  };
+  const response = await sendForm(waitlistForm.action, formData);
+  // So far, always a Bad Request
+	return;
```
Comment (lesleyjanenorton) [inline static/js/app.js:164]: @groovecoder Was there some incantation for handling the alias forwarding preference form this way? I can't remember.

File: static/js/app.js [hunk 5]
```diff
@@ -39,6 +39,7 @@ async function sendForm(formAction, formData) {
 			headers: {
 				"Content-Type": "application/json",
 				"X-Requested-With": "XMLHttpRequest",
+        "X-CSRFToken": formData.csrfmiddlewaretoken,
```
Comment (groovecoder) [inline static/js/app.js:42]: Oh, this should probably be made more generic?

**Comment by groovecoder** [comment]
Added code that fixes the `fetch` POST to the waitlist. Also added code to check if the user is already on the wait-list and to show them a message.

**Comment by groovecoder** [comment]
Has some conflicts in `style.css` and `app.js`

File: privaterelay/signals.py [hunk 6]
```diff
@@ -31,31 +31,47 @@ def invitations_only(sender, **kwargs):
...
+            )
+            kwargs['request'].session['already_on_waitlist'] = (
+                waitlist_invite.count() > 0
+            )
+        else:
+            kwargs['request'].session['waitlist_open'] = False
+
+        # If we're not doing token-based invites; reject immediately
+        if settings.ALPHA_INVITE_TOKEN:
+            # Token inviations are subject to max accounts limit
```
Comment (jens1o) [inline privaterelay/signals.py:52]:
```suggestion
            # Token invitations are subject to max accounts limit
```

**Pull request merged**